content/posts/01-gsoc-2020.rst

```diff
@@ -0,0 +1,198 @@
+refactoring to the HID dissector and changed the HID report descriptor
+annotations to have the data look like as shown in the specification_. I also
+added some missing HID usage pages.
+
+This allowed me to get to know a bit more about the Wireshark internals, as well
+as how the HID dissector was implemented. I think this step was incredibly
+helpful.
+
+
+The work
+========
+
+Before I started working on the HID report descriptor parser, I spent a few days
+with the HID specification. This allowed me to familiarize myself more with the
+intricacies of the of HID and the report descriptor format. It helped a lot
+designing the data structures and the parser. Even with this preparation, I did
+not get it right at first, but I got it close enough that it was easy to fix.
+
+After having the parser implemented, I started working on the data annotations.
+Along the way, I natuarally tweaked the parser to better adapt to our use-case.
+The step of writting the annotations also presented its issues.
+Wireshark operates data at a byte level, but HID operates at a bit level. There
+were already some internal ways to try to get around this issue, but they needed
+to be expanded to properly cater to the needs of HID. Wireshark has encontered
+some protocols needed to dynamically annotate data on a bit level, but their
+requirements were simpler, they mostly wanted to annotate a fixed data type.
+HID, however, can have long data fields, so our main question here was how to
+annotate byte arrays that are bit limited? The same approach used in other types
+of data couldn't be applied here.
+
+Overcoming these issues, I think we got some nice end results.
+
+
+.. figure:: ../../images/posts/01/mouse-dissection.png
+
+    Image 1 - Mouse data dissection
+
+
+.. figure:: ../../images/posts/01/keyboard-dissection.png
+
+    Image 2 - Keyboard data dissection
+
+
+.. figure:: ../../images/posts/01/joystick-dissection.png
+
+    Image 3 - Joystick data dissection
+
+
+.. figure:: ../../images/posts/01/vendor-dissection.png
+
+    Image 4 - Vendor data dissection
+
+
+As you can see, annotations for some fields are still missing. I will keep
+adding more fields as time allows. What is important here is that adding
+support for more fields is easy.
+
+
+Gotchas
+=======
+
+Missing descriptors
+-------------------
+
+To dissect the HID data we need access to the report descriptor, which tells us
+how the reports are structured. This is also the case of for eg. the USB
+endpoint descriptors, we need to look at the endpoint descriptors to understand
+the endpoints. But there is an issue, we are dissecting from a capture, either
+live or from a file, we may not have access to the device to fetch this
+descriptors. Because of this, we need the packets that ask the device for this
+information to be included in the capture.
+Fortunately, libpcap_ asks the device for the USB endpoint descriptors when you
```

Comment:
lipcap asks for the DEVICE descriptor when you start capture. If you have recent enough (https://github.com/the-tcpdump-group/libpcap/commit/ab403787c2b0a4850b1c7791bb7dc190ef83b35c) libpcap version, it will also ask for CONFIGURATION descriptor.